src/Castle.Windsor/MicroKernel/Arguments.cs

```diff
@@ -19,208 +19,240 @@ namespace Castle.MicroKernel
+			get => dictionary[key];
+			set
+			{
+				EnsureWritable();
+				dictionary[key] = value;
+			}
 		}
 
-		public ICollection Keys
-		{
-			get { return arguments.Keys; }
-		}
+		public int Count => dictionary.Count;
 
-		public ICollection Values
-		{
-			get { return arguments.Values; }
-		}
+		public ICollection Keys => dictionary.Keys;
 
-		bool ICollection.IsSynchronized
-		{
-			get { return arguments.IsSynchronized; }
-		}
+		public ICollection Values => dictionary.Values;
+
+		bool ICollection.IsSynchronized => dictionary.IsSynchronized;
+
+		object ICollection.SyncRoot => dictionary.SyncRoot;
+
+		bool IDictionary.IsFixedSize => dictionary.IsFixedSize;
+
+		bool IDictionary.IsReadOnly => isReadOnly;
+
+		public static readonly Arguments Empty = new Arguments { isReadOnly = true };
 
-		object ICollection.SyncRoot
+		public void Add(object key, object value)
 		{
-			get { return arguments.SyncRoot; }
+			EnsureWritable();
+			dictionary.Add(key, value);
 		}
 
-		bool IDictionary.IsFixedSize
+		public void Clear()
 		{
-			get { return arguments.IsFixedSize; }
+			EnsureWritable();
+			dictionary.Clear();
 		}
 
-		bool IDictionary.IsReadOnly
+		public Arguments Clone()
 		{
-			get { return arguments.IsReadOnly; }
+			return new Arguments(dictionary);
 		}
 
-		public void Add(object key, object value)
+		void ICollection.CopyTo(Array array, int index)
 		{
-			arguments.Add(key, value);
+			dictionary.CopyTo(array, index);
 		}
 
-		public void Clear()
+		public bool Contains(object key)
 		{
-			arguments.Clear();
+			return dictionary.Contains(key);
 		}
 
-		public bool Contains(object key)
+		IEnumerator IEnumerable.GetEnumerator()
 		{
-			return arguments.Contains(key);
+			return GetEnumerator();
 		}
 
 		public IDictionaryEnumerator GetEnumerator()
 		{
-			return arguments.GetEnumerator();
+			return dictionary.GetEnumerator();
 		}
 
-		public void Remove(object key)
+		/// <summary>
+		/// Inserts a named argument. If the argument already exists it will be overwritten.
+		/// </summary>
+		public virtual Arguments Insert(string key, object value)
 		{
-			arguments.Remove(key);
+			this[key] = value;
+			return this;
 		}
 
-		public Arguments Clone()
+		/// <summary>
+		/// Inserts a set of named and/or typed arguments. If an argument already exists it will be overwritten.
+		/// </summary>
+		public virtual Arguments Insert(IDictionary values)
 		{
-			var dictionary = arguments as Dictionary<object, object>;
-			if (dictionary != null)
+			foreach (DictionaryEntry entry in values)
 			{
-				var comparerExtended = dictionary.Comparer as ArgumentsComparerExtended;
-				if (comparerExtended != null)
+				if (entry.Key is string || entry.Key is Type)
+				{
+					this[entry.Key] = entry.Value;
+				}
+				else
 				{
-					return new Arguments(arguments, comparerExtended.CustomComparers);
+					throw new ArgumentException($"The argument '{entry.Key}' should be of type string or System.Type");
```

Comment:
The BCL ends exception messages with periods, and I see this [recommended](https://docs.microsoft.com/dotnet/standard/exceptions/best-practices-for-exceptions#use-grammatically-correct-error-messages) too.